sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala

```diff
@@ -246,85 +260,238 @@ object SpecifiedWindowFrame {
+ * offset -2, will get the value of x 2 rows back in the partition.
+ */
+abstract class OffsetWindowFunction
+  extends Expression with WindowFunction with Unevaluable with ImplicitCastInputTypes {
+  val input: Expression
+  val default: Expression
+  val offset: Expression
+  val offsetSign: Int
+
+  override def children: Seq[Expression] = Seq(input, offset, default)
 
-  def prepareInputParameters(input: InternalRow): AnyRef
+  override def foldable: Boolean = input.foldable && (default == null || default.foldable)
 
-  def update(input: AnyRef): Unit
+  override def nullable: Boolean = input.nullable && (default == null || default.nullable)
 
-  def batchUpdate(inputs: Array[AnyRef]): Unit
+  override lazy val frame = {
+    // This will be triggered by the Analyzer.
+    val offsetValue = offset.eval() match {
+      case o: Int => o
+      case x => throw new AnalysisException(
+        s"Offset expression must be a foldable integer expression: $x")
+    }
+    val boundary = ValueFollowing(offsetSign * offsetValue)
+    SpecifiedWindowFrame(RowFrame, boundary, boundary)
+  }
 
-  def evaluate(): Unit
+  override def dataType: DataType = input.dataType
 
-  def get(index: Int): Any
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(AnyDataType, IntegerType, TypeCollection(input.dataType, NullType))
 
-  def newInstance(): WindowFunction
+  override def toString: String = s"$prettyName($input, $offset, $default)"
 }
 
-case class UnresolvedWindowFunction(
-    name: String,
-    children: Seq[Expression])
-  extends Expression with WindowFunction with Unevaluable {
+case class Lead(input: Expression, offset: Expression, default: Expression)
+    extends OffsetWindowFunction {
 
-  override def dataType: DataType = throw new UnresolvedException(this, "dataType")
-  override def foldable: Boolean = throw new UnresolvedException(this, "foldable")
-  override def nullable: Boolean = throw new UnresolvedException(this, "nullable")
-  override lazy val resolved = false
+  def this(input: Expression, offset: Expression) = this(input, offset, Literal(null))
 
-  override def init(): Unit = throw new UnresolvedException(this, "init")
-  override def reset(): Unit = throw new UnresolvedException(this, "reset")
-  override def prepareInputParameters(input: InternalRow): AnyRef =
-    throw new UnresolvedException(this, "prepareInputParameters")
-  override def update(input: AnyRef): Unit = throw new UnresolvedException(this, "update")
-  override def batchUpdate(inputs: Array[AnyRef]): Unit =
-    throw new UnresolvedException(this, "batchUpdate")
-  override def evaluate(): Unit = throw new UnresolvedException(this, "evaluate")
-  override def get(index: Int): Any = throw new UnresolvedException(this, "get")
+  def this(input: Expression) = this(input, Literal(1))
 
-  override def toString: String = s"'$name(${children.mkString(",")})"
+  def this() = this(Literal(null))
```

Comment:
This is only a default constructor for serialization purposes. It shouldn't really be used by anyone. The data type of this expression will be ```NullType```, ```Literal.apply``` takes care of that.